server/server/src/cold_chain/temperature_log.rs

```diff
@@ -0,0 +1,143 @@
+    sensor_id: String,
+    #[serde(rename = "temperatureBreachId")]
+    pub temperature_breach_id: Option<String>,
+}
+
+pub async fn put_logs(
+    request: HttpRequest,
+    service_provider: Data<ServiceProvider>,
+    auth_data: Data<AuthData>,
+    logs: web::Json<Vec<TemperatureLog>>,
+) -> HttpResponse {
+    let store_id = match validate_request(request, &service_provider, &auth_data) {
+        Ok((_user, store_id)) => store_id,
+        Err(error) => {
+            let formatted_error = format!("{:#?}", error);
+            return HttpResponse::Unauthorized().body(formatted_error);
+        }
+    };
+    let logs = logs.into_inner();
+    if !validate_input(&logs) {
+        return HttpResponse::BadRequest()
+            .body("Expecting a body with the array of temperature logs");
+    };
+
+    let results = match upsert_temperature_logs(service_provider, store_id, &logs).await {
+        Ok(response) => response,
+        Err(error) => return HttpResponse::InternalServerError().body(format!("{:#?}", error)),
+    };
+
+    HttpResponse::Ok()
+        .append_header(header::ContentType(mime::APPLICATION_JSON))
+        .json(&results)
+}
+
+fn validate_input(logs: &Vec<TemperatureLog>) -> bool {
+    logs.iter().all(|log| validate_log(log))
+}
+
+fn validate_log(log: &TemperatureLog) -> bool {
+    if log.timestamp < 1 {
+        return false;
+    }
+    true
+}
+
+async fn upsert_temperature_logs(
+    service_provider: Data<ServiceProvider>,
+    store_id: String,
+    logs: &Vec<TemperatureLog>,
+) -> Result<Vec<Result<repository::TemperatureLog, String>>, RepositoryError> {
+    let mut ctx = service_provider.basic_context()?;
+    ctx.store_id = store_id;
+    let service = &service_provider.temperature_log_service;
+
+    let results = logs
+        .iter()
+        .map(|log| {
+            let id = log.id.clone();
+            let datetime = match NaiveDateTime::from_timestamp_opt(log.timestamp, 0) {
+                Some(datetime) => datetime,
+                None => {
+                    error!(
+                        "Unable to parse timestamp for log {}: {}",
+                        log.id, log.timestamp
+                    );
+                    return Err(format!(
+                        "Unable to parse timestamp for log {}: {}",
+                        log.id, log.timestamp
+                    ));
+                }
+            };
+            match service.get_temperature_log(&ctx, log.id.clone()) {
+                Ok(_) => {
+                    let log = UpdateTemperatureLog {
+                        id: id.clone(),
+                        datetime,
+                        location_id: None,
+                        sensor_id: log.sensor_id.clone(),
+                        temperature: log.temperature,
+                        temperature_breach_id: log.temperature_breach_id.clone(),
+                    };
+                    match service.update_temperature_log(&ctx, log) {
```

Comment:
Why would temperature log ever be updated ?

Reply:
yeah, it probably won't ever happen. I went with the upsert because that's what the existing API is doing - it checks for an existing record (based on ID) and updates if it finds one and inserts if not.

It is a little odd, but it easy enough to implement so I went that way. Could have it throw an error instead, write an error to the log and skip the record. Would you prefer that?